OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,286 @@
+holding metadata. Data contains an independent column and a set of dependent 
+columns. The type of the independent column can be configured using ETX (
+template param). The type of the dependent columns, which together form a matrix
+, can be configured using ETY (template param). Independent and Dependent 
+columns can contain metadata. DataTable_ as a whole can contain metadata.     */
+template<typename ETX = double, typename ETY = SimTK::Real>
+class DataTable_ : public AbstractDataTable {
+public:
+    using RowVector    = SimTK::RowVector_<ETY>;
+    using ColumnVector = SimTK::Vector_<ETY>;
+
+    DataTable_()                             = default;
+    DataTable_(const DataTable_&)            = default;
+    DataTable_(DataTable_&&)                 = default;
+    DataTable_& operator=(const DataTable_&) = default;
+    DataTable_& operator=(DataTable_&&)      = default;
+    ~DataTable_()                            = default;
+
+    std::unique_ptr<AbstractDataTable> clone() const override {
+        return std::unique_ptr<AbstractDataTable>{new DataTable_{*this}};
+    }
+
+    /** Append row to the DataTable_.                                         */
+    void appendRow(const ETX& indRow, const RowVector& depRow) {
+        validateAppendRow(indRow, depRow);
+
+        _indData.push_back(indRow);
+
+        if(_depData.nrow() == 0 || _depData.ncol() == 0) {
+            try {
+                auto& labels = 
+                    _dependentsMetaData.getValueArrayForKey("labels");
+                if(depRow.size() != labels.size())
+                    throw Exception{"Number of columns in the input row does "
+                            "not match the number of labels."};
+            } catch(std::out_of_range&) {
+                // No operation.
+            }
+            _depData.resize(1, depRow.size());
+        } else
+            _depData.resizeKeep(_depData.nrow() + 1, _depData.ncol());
+            
+        _depData.updRow(_depData.nrow() - 1) = depRow;
+    }
+
+    /** Get row at index.                                                     */
+    RowVector getRowAtIndex(size_t index) const {
+        return _depData.row(index);
+    }
+
+    /** Get row corresponding to the given entry in the independent column.   */
+    RowVector getRow(const ETX& ind) const {
+        auto iter = std::find(_indData.cbegin(), _indData.cend(), ind);
+
+        if(iter == _indData.cend())
+            throw Exception{"Independent column has no entry with value " +
+                    std::to_string(ind)};
+
+        return _depData.row(std::distance(_indData.cbegin(), iter));
+    }
+
+    /** Update row at index.                                                  */
+    RowVector updRowAtIndex(size_t index) {
+        return _depData.updRow(index);
+    }
+
+    /** Update row corresponding to the given entry in the independent column.*/
+    RowVector updRow(const ETX& ind) {
```

Comment:
I don't think this works. `Matrix::updData()` returns a `RowVectorView` but then that is being copied to a `RowVector` on return. These are identical types except that `RowVector` has deep copy semantics while `RowVectorView` has shallow copy semantics. So I think the function return will get a copy of the data here.